test/run_clippy.sh

```diff
@@ -0,0 +1,138 @@
+                "clippy::new_without_default_derive")
+
+ # Known failing lints we want to receive warnings for, but not fail the build
+lints_to_fix=("clippy::cyclomatic_complexity" \
+               "clippy::large_enum_variant" \
+               "clippy::len_without_is_empty" \
+               "clippy::needless_pass_by_value" \
+               "clippy::needless_return" \
+               "clippy::question_mark" \
+               "clippy::redundant_field_names" \
+               "clippy::too_many_arguments" \
+               "clippy::trivially_copy_pass_by_ref" \
+               "clippy::wrong_self_convention")
+
+ # Lints we don't expect to have in our code at all and want to avoid adding
+# even at the cost of failing the build
+denied_lints=("clippy::assign_op_pattern" \
+               "clippy::blacklisted_name" \
+               "clippy::block_in_if_condition_stmt" \
+               "clippy::bool_comparison" \
+               "clippy::cast_lossless" \
+               "clippy::clone_on_copy" \
+               "clippy::cmp_owned" \
+               "clippy::collapsible_if" \
+               "clippy::const_static_lifetime" \
+               "clippy::correctness" \
+               "clippy::deref_addrof" \
+               "clippy::expect_fun_call" \
+               "clippy::for_kv_map" \
+               "clippy::get_unwrap" \
+               "clippy::identity_conversion" \
+               "clippy::if_let_some_result" \
+               "clippy::len_zero" \
+               "clippy::let_and_return" \
+               "clippy::let_unit_value" \
+               "clippy::map_clone" \
+               "clippy::match_bool" \
+               "clippy::match_ref_pats" \
+               "clippy::needless_bool" \
+               "clippy::needless_collect" \
+               "clippy::needless_range_loop" \
+               "clippy::ok_expect" \
+               "clippy::op_ref" \
+               "clippy::option_map_unit_fn" \
+               "clippy::or_fun_call" \
+               "clippy::println_empty_string" \
+               "clippy::ptr_arg" \
+               "clippy::redundant_closure" \
+               "clippy::redundant_pattern_matching" \
+               "clippy::single_char_pattern" \
+               "clippy::single_match" \
+               "clippy::string_lit_as_bytes" \
+               "clippy::toplevel_ref_arg" \
+               "clippy::unit_arg" \
+               "clippy::unnecessary_operation" \
+               "clippy::unreadable_literal" \
+               "clippy::unused_label" \
+               "clippy::unused_unit" \
+               "clippy::useless_asref" \
+               "clippy::useless_format" \
+               "clippy::useless_let_if_seq" \
+               "clippy::useless_vec" \
+               "clippy::write_with_newline" \
+               "renamed_and_removed_lints")
+
+ # stick together unexamined lints
+for unexamined_lint in "${unexamined_lints[@]}"
+do
+  [[ -n $unexamined_lint ]] && rules_string+="-A ${unexamined_lint} "
```

Comment:
There's not need to check for a zero-length string; just don't add such strings to your array in the first place.

Also, instead of building up `rules_string`, let's build up an array: `clippy_args`:
```suggestion
  clippy_args+=(-A "${unexamined_lint}")
```

Reply:
What would be even *better*, would be a function to abstract the building up of the `clippy_args` array with the different flag values:
```bash
clippy_args=()

add_lints_to_clippy_args() {
  flag=$1
  shift
  for lint
  do
    clippy_args+=("$flag" "${lint}")
  done
}

add_lints_to_clippy_args -A "${unexamined_lints[@]}"
add_lints_to_clippy_args -A "${allowed_lints[@]}"
add_lints_to_clippy_args -W "${lints_to_fix[@]}"
add_lints_to_clippy_args -D "${denied_lints[@]}"
```

Reply:
And what would be *better **yet*** would be instead of defining the lints in the different categories inline in the script, read them into arrays from separate files (e.g. lints_to_fix.txt) using `mapfile`. That way we can use the same source of lints for both this build kite script and for local execution.

That's assuming we can't make this script work for both, in which case, defining the lints here may be fine. I just don't want to have to keep this script and the `Makefile` in sync.